### PR TITLE
Pull-Request-Squash-Commit

### PR DESCRIPTION
**Pull request by Sarmad93**
Implementation of  #1236 opened by @ryangribble for [Preview Support for Pull Request Merge Api](https://developer.github.com/changes/#--preview-squash-support-for-the-pull-request-merge-api) recently introduced by github.

### CODE REVIEW

File: Octokit.Tests/Reactive/ObservablePullRequestsClientTests.cs [hunk 1]
```diff
@@ -225,7 +225,7 @@ public class TheMergeMethod
             [Fact]
             public void MergesPullRequest()
             {
-                var mergePullRequest = new MergePullRequest { CommitMessage = "fake commit message" };
+                var mergePullRequest = new MergePullRequest { CommitMessage = "fake commit message", CommitTitle = "fake title" };
```
Comment (shiftkey) [inline Octokit.Tests/Reactive/ObservablePullRequestsClientTests.cs]: Is this a necessary change?

Comment (Sarmad93) [inline Octokit.Tests/Reactive/ObservablePullRequestsClientTests.cs]: No this is not necessary. but i thought i should also make changes in ObservablePullRequestsClientTests.cs .I wasn't sure that i should add property here or leave it as it was.. what do you say should i remove this?

Comment (shiftkey) [inline Octokit.Tests/Reactive/ObservablePullRequestsClientTests.cs]: It looks like we're not asserting anything about this object, so I think we should revert this change. :cool:?

Comment (Sarmad93) [inline Octokit.Tests/Reactive/ObservablePullRequestsClientTests.cs]: yes you are right..i should revert it now. :+1:

**Comment by Sarmad93** [comment]
oops look like i did some thing wrong.

File: Octokit/Helpers/AcceptHeaders.cs [hunk 2]
```diff
@@ -15,5 +15,7 @@ public static class AcceptHeaders
         public const string StarCreationTimestamps = "application/vnd.github.v3.star+json";
 
         public const string CommitReferenceSha1Preview = "application/vnd.github.chitauri-preview+sha";
+
+        public const string SquashCommitPreview = "application/vnd.github.polaris-preview+json";
```
Comment (shiftkey) [inline Octokit/Helpers/AcceptHeaders.cs]: I'm not seeing where we use this header in the client itself. This should do something inside `PullRequestClient.Merge(..., MergePullRequest)`

File: Octokit.Tests.Integration/Clients/PullRequestsClientTests.cs [hunk 3]
```diff
@@ -225,6 +226,20 @@ public PullRequestsClientTests()
...
+    {
+        await CreateTheWorld();
+
+        var newPullRequest = new NewPullRequest("a pull request", branchName, "master");
+        var pullRequest = await _fixture.Create(Helper.UserName, _context.RepositoryName, newPullRequest);
+
+        var merge = new MergePullRequest { CommitMessage = "fake message", CommitTitle = "fake title", Squash = true };
+        var result = await _fixture.Merge(Helper.UserName, _context.RepositoryName, pullRequest.Number, merge);
+
+        Assert.True(result.Merged);
```
Comment (shiftkey) [inline Octokit.Tests.Integration/Clients/PullRequestsClientTests.cs]: Could we assert something here about the commit title or message versus what we actually committed, to indicate this was squashed?

**Comment by shiftkey** [comment]
@Sarmad93 thanks for starting this off. I added a couple of comments to help us complete this feature and verify we're doing everything right...

File: Octokit.Tests.Integration/Clients/PullRequestsClientTests.cs (hunk 3 above)
Comment (Sarmad93) [inline Octokit.Tests.Integration/Clients/PullRequestsClientTests.cs]: sure i will add assert statement here. :smiley:

**Comment by Sarmad93** [comment]
@shiftkey thanks for guiding me. Sure we wil complete this one :+1:

File: Octokit.Tests.Integration/Clients/PullRequestsClientTests.cs (hunk 3 above)
Comment (shiftkey) [inline Octokit.Tests.Integration/Clients/PullRequestsClientTests.cs]: One suggestion - look at the commit:

```
var commit = await _github.Repository.Commit.Get(_context.RepositoryOwner, _context.RepositoryName, result.Sha);
// TODO: something interesting goes here
...

Comment (Sarmad93) [inline Octokit.Tests.Integration/Clients/PullRequestsClientTests.cs]: :+1:

**Comment by Sarmad93** [comment]
hi @shiftkey i am doing changes in integration test. when i ran build i saw that integration test were actually not running(skipping) because of some error
>OCTOKIT_GITHUBUSERNAME and OCTOKIT_GITHUBPASSWORD environment variables are not set.


how can i make sure my test are passing? do i need to make a separate account for this?
thanks.

**Comment by ryangribble** [comment]
> how can i make sure my test are passing? do i need to make a separate testing account for this?
thanks.

@Sarmad93 yeah you should create a test account because the integration tests do things like create repos and other things.  Generally they try to clean up after themselves where possible, but definitely you want a separate test account to do this under.

Refer to the integration test sectino in [Contributors Guide](https://github.com/octokit/octokit.net/blob/master/CONTRIBUTING.md#slow-tests) for information about how to run the tests (essentially there's a powershell script to help you define those environment variables)

**Comment by Sarmad93** [comment]
@ryangribble  i will test these against testing account but i am pushing some changes could you please review these especially `header` functionality? i am not sure how this header parameter functionality works. i am stuck at this one need little help.

**Comment by Sarmad93** [comment]
sorry for incomplete commits this smartgit github client really has some weird behavior :angry: doesn't always add all files to staging area.

**Comment by ryangribble** [comment]
Im not in a position to pull your code down and run it myself, but just eyeballing your changes, yes that looks correct in specifying the accept header on the `ApiConnection.Put<T>` call.  

However on the unit test mock verification you shouldnt use `AcceptHeaders.xxxxx` there (instead hardcode the expected header string here) as we want the unit test to independently verify the implementation code, so we dont want to "share" that `AcceptHeaders` helper between implementation and tests.  Same way we manually check the Url that is called rather than use the `ApiUrl` helper method :grinning:

**Comment by Sarmad93** [comment]
@ryangribble thanks for the code review :satisfied: yes i need to test these on testing account now.
Regarding `AcceptHeaders.xxxxx` in unit test, you explained very well i need to pass header as string in parameter :+1:

**Comment by Sarmad93** [comment]
i have configured my test account for integration test by running script in `scripts\ configure-integration-tests.ps1` now i am getting some other errors from ` Octokit\GitHubClient.cs` . i thinks my username and password is not updating from script. how can i resolve this?
capture11

**Comment by Sarmad93** [comment]
i am stuck in this testing phase :disappointed:

**Comment by shiftkey** [comment]
@Sarmad93 looking at those errors, I think something wasn't set correctly. You can run the same script again to view and update the values that the tests need.

**Comment by shiftkey** [comment]
Just putting this back to `WIP` until we can confirm the integration test setup script is correct.

**Comment by ryangribble** [comment]
Hey @Sarmad93, I'd definitely like to ensure these new GitHub features make it into the next octokit release so if time is an issue for you to get back onto this (totally understandable!) just let me know if you want someone to pick it up from here :grinning:

**Comment by Sarmad93** [comment]
@ryangribble yes i have been busy for last days and last time i remember i was stuck on integration test setup and i wasn't sure where to add parameter(header) in post request so i stopped working on this :tired_face: got  tired. you can assign this to someone else. Really sorry i could not made it.

**Comment by ryangribble** [comment]
If you want to keep working on it you're more than welcome to! 

It looks like you were at the point where we needed you to confirm your integration test settings...  Easiest way is just to close visual studio, rerun the powershell script and ensure they are all set correctly, then have another go at running integration tests

**Comment by Sarmad93** [comment]
i made a test account and set the script against testing account but after setting up the script i was getting these errors. why these test are failing? most of the test in `PullRequestsClientTests.cs`are failing..they made me annoyed  :smiley: an exception is thrown from connection.cs file. something is not setting up or there is some thing wrong in response :confused:  
octokit2

**Comment by Sarmad93** [comment]
@ryangribble also when i build project in visual studio 2015 i get this error also 
octokit4

**Comment by shiftkey** [comment]
@Sarmad93 some of the integration tests depend on the account being in a certain state - I've tried to keep it as runnable as possible for everyone, but differences do creep in which means they can fail.

If you want to focus on just your test, I'd suggest running it from in Visual Studio.

File: SolutionInfo.cs [hunk 4]
```diff
@@ -6,10 +6,8 @@
 [assembly: AssemblyVersionAttribute("0.19.0")]
 [assembly: AssemblyFileVersionAttribute("0.19.0")]
 [assembly: ComVisibleAttribute(false)]
-namespace System
-{
-    internal static class AssemblyVersionInformation
-    {
+namespace System {
+    internal static class AssemblyVersionInformation {
```
Comment (devkhan) [inline SolutionInfo.cs]: This is unintended, right?

Comment (shiftkey) [inline SolutionInfo.cs]: I think this is a side-effect of `.\build FormatCode` - I'm tracking this in https://github.com/shiftkey/Octokit.CodeFormatter/issues/2

**Comment by Sarmad93** [comment]
unit tests + convention tests build is passing locally but not here? :confused:

**Comment by devkhan** [comment]
@Sarmad93 Your has conflicts with `master`, and that's why the CI builds are not running. Try to rebase your branch.

**Comment by Sarmad93** [comment]
@devkhan  still not building?

**Comment by devkhan** [comment]
What did you do exactly?

**Comment by Sarmad93** [comment]
@devkhan i did only this `git rebase master` and all that happened as you can see above :D

**Comment by devkhan** [comment]
Did you updated you master from upstream first? You have to do that as that will fetch all the latest commits made in the main repo's master. Then you have to do the following to resolve conflicts:

```bash
git checkout <your_branch>
git rebase master
# Resolve any conflicts that may arise manually
# Then marked them resolved by:
git add .
# Then continue rebasing:
git rebase --continue
# Finally do a force push:
git push --force origin <your_branch>
```

Let me what happens after doing the above steps. :+1:

**Comment by Sarmad93** [comment]
@devkhan on step 4(`git rebase --continue`) i am getting this. :confused: 
capture00

**Comment by devkhan** [comment]
Did you resolved the conflicting files and then did `rebase --continue`?

Remember, you have to keep performing `rebase --continue`, conflict resolution and `git add .` till all the commits are rebased successfully.

**Comment by Sarmad93** [comment]
@devkhan  i think something terrible has happened to my source tree. i added and resolved all conflicts iteratively but still `appveyor` is showing :x:

**Comment by devkhan** [comment]
@Sarmad93 I tried rebasing your branch on my own in my fork. As you can see below:

image

Unfortunately, there are some changes also in the commtis made by others which are already merged in `master`. So, what I would suggest is, that you note down the `sha`s of the commits you have made, the ones listed in the image (but in your fork). Then do the following:

```git
git checkout Squash_Commit
git checkout -b sc-temp # For backup
git checkout Squash_Commit
git reset --hard HEAD~200 # Go back well enough in history
git pull upstream master # Update from octokit's master
git cherry-pick <sha> # Do for all shas noted down before
git push --force origin Squash_Commit
```

Let me know what happens then.

**Comment by devkhan** [comment]
And if there are any conflicts while `cherry-picking`, resolve them and commit too.

**Comment by Sarmad93** [comment]
@devkhan sir first i would to thank you so much for your time and efforts :+1: i think this pr has gone to some fatal condition.. appveyor build is generating some strange errors :laughing:

**Comment by Sarmad93** [comment]
@devkhan i have to redo this in a single commit. maybe it will solve the problem.

**Comment by devkhan** [comment]
The reasons of those errors are the diffs which are inserted into the files on conflicts. I also think that a new PR may be better or if the conversation here is relevant, you can go actually delete your branch locally and then recreate it from the latest upstream master, and then re-do your work on top of it. And finally, do a force push.

**Comment by Sarmad93** [comment]
i want to get back on this. i have free time for a while :smile: . I need a little help.
regarding passing the `Header` in put request i think i need to pass it in `PullRequestClient.Merge` . This merge method returns `PullRequestMerge` model object. In return statement the put method takes 4 input arguments. The one which confuses me the 3rd parameter `string twoFactorAuthenticationCode` Should i pass null here? or dummy text or i should update the position of parameters in overloaded `put` in `IConnection` class so that make `Accept Header` third input parameter and leave `twoFactorAuthenticationCode` without initializing it.

**Comment by shiftkey** [comment]
@Sarmad93 

> The one which confuses me the 3rd parameter string `twoFactorAuthenticationCode` Should i pass null here? 

You can specify `null` fine here.

> or i should update the position of parameters in overloaded put in IConnection class so that make Accept Header third input parameter and leave twoFactorAuthenticationCode without initializing it.

Don't reorder the parameters as part of this PR. I'm not really sure why `twoFactorAuthenticationCode` is so pervasive in the `IConnection` signatures but let's avoid that for now...

**Comment by Sarmad93** [comment]
@shiftkey could you please review changes i have made especially for `Accept` header?
Integration tests are not working for me. so i have tested it on creating new project and using access token. it works fine for me and squashes more than one commit into a single commit. kindly if you pull this code and run the integration test by yourself so that it could be verified that it is working as it  intended to be :smile:

File: Octokit.Tests/OctoKit.Tests-NetCore45.csproj [hunk 5]
```diff
@@ -184,6 +184,10 @@
       <Project>{c8bc13b6-3fa3-4716-827d-e7706f976fe1}</Project>
       <Name>Octokit-NetCore45</Name>
     </ProjectReference>
+    <ProjectReference Include="..\Octokit\Octokit.csproj">
```
Comment (shiftkey) [inline Octokit.Tests/OctoKit.Tests-NetCore45.csproj]: Not sure how this crept in, but this project reference shouldn't be needed :fire: :fire: :fire:

**Comment by Sarmad93** [comment]
Really sorry for that... i accidentally add all the files to staging area without observing closely...i will fix this and push new changes..sorry again if this made you angry..

**Comment by shiftkey** [comment]
@Sarmad93 I'm not angry, I just like to delete code whenever possible :fire:

**Comment by shiftkey** [comment]
@Sarmad93 I'm happy with that integration test - let's wrap this up and get it merged!

**Comment by Sarmad93** [comment]
@shiftkey done.

**Comment by shiftkey** [comment]
@Sarmad93 excellent, thanks!

**Comment by Sarmad93** [comment]
@shiftkey @ryangribble as always thanks for your code review and guidance :+1: you guyz helped me to finish it to end.
@devkhan sir you taught me how to resolve merge/rebase conflict rightly,appreciate your efforts.

**Comment by devkhan** [comment]
@Sarmad93 Glad I could help. You don't need to call me "Sir", we're all here to help each other. And I'm happy this PR is finally merged. Keep up the good work. :+1:
